Tools/MiniBrowser/win/WebKitBrowserWindow.cpp

```diff
@@ -573,6 +578,64 @@ bool WebKitBrowserWindow::canTrustServerCertificate(WKProtectionSpaceRef protect
     return false;
 }
 
+void WebKitBrowserWindow::navigationActionDidBecomeDownload(WKPageRef, WKNavigationActionRef, WKDownloadRef download, const void* clientInfo)
+{
+    setDownloadClient(download, clientInfo);
+}
+
+void WebKitBrowserWindow::navigationResponseDidBecomeDownload(WKPageRef, WKNavigationResponseRef, WKDownloadRef download, const void* clientInfo)
+{
+    setDownloadClient(download, clientInfo);
+}
+
+void WebKitBrowserWindow::contextMenuDidCreateDownload(WKPageRef, WKDownloadRef download, const void* clientInfo)
+{
+    setDownloadClient(download, clientInfo);
+}
+
+void WebKitBrowserWindow::setDownloadClient(WKDownloadRef download, const void* clientInfo)
+{
+    WKDownloadClientV0 client = { };
+    client.base = { 0, clientInfo };
+    client.decideDestinationWithResponse = downloadDecideDestinationWithResponse;
+    client.didFinish = downloadDidFinish;
+    client.didFailWithError = downloadDidFailWithError;
+
+    WKDownloadSetClient(download, &client.base);
+}
+
+WKStringRef WebKitBrowserWindow::downloadDecideDestinationWithResponse(WKDownloadRef download, WKURLResponseRef response, WKStringRef suggestedFilename, const void* clientInfo)
+{
+    auto filename = createString(suggestedFilename);
+    if (filename.empty()) {
+        auto urlRef = adoptWK(WKURLResponseCopyURL(response));
+        auto hostRef = adoptWK(WKURLCopyHostName(urlRef.get()));
+        filename = createString(hostRef.get());
+    }
+
+    std::wstring folderPath;
+    getKnownFolderPath(FOLDERID_Downloads, folderPath);
+
+    return createWKString(folderPath + L"\\" + filename).leakRef();
```

Comment:
What will happen if the file already exists? Overwrite without a warning? It seems too dangerous.

Reply:
The file is not overwritten because WebKit::AllowOverwrite::No is specified. Instead, it raises a CURLE_WRITE_ERROR error if the file already exists.

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/UIProcess/API/C/WKDownloadRef.cpp#L114

However, NetworkDataTaskCurl did not include overwrite processing, so I added setPendingDownloadLocation() and FileSystem::openFile() processing referring to NetworkDataTaskDataURL.

Reply:
Does any message boxes appear in this case? What does it say?

Reply:
<img width="248" alt="ErrorDialog" src="https://github.com/WebKit/WebKit/assets/113365892/3cb6a5c2-8e4c-47d8-88bb-9378fea6faca">

Reply:
<img width="248" alt="ErrorDialog" src="https://github.com/WebKit/WebKit/assets/113365892/8d18bd8b-8eea-449e-813a-212c02ed76cc">